File: frontend/src/api/learner.ts
const BASE = 'http://localhost:8000';

export interface ConceptMastery {
  concept_name: string;
  mastery_score: number;
  total_exposures: number;
  last_tested: string | null;
}

export interface Misconception {
  concept_name: string;
  pattern: string;
  severity: 'high' | 'medium' | 'low';
  repair_strategy: string;
}

export interface LearnerStats {
  concepts: ConceptMastery[];
  misconceptions: Misconception[];
  weak_count: number;
}

export async function fetchLearnerStats(token: string): Promise<LearnerStats | null> {
  try {
    const res = await fetch(`${BASE}/learner/stats?token=${encodeURIComponent(token)}`);
    if (!res.ok) return null;
    return (await res.json()) as LearnerStats;
  } catch {
    return null;
  }
}

File: backend/routers/learner.py
from fastapi import APIRouter, HTTPException, Query
from ..auth.utils import decode_token_active
from ..memory import longterm_memory

router = APIRouter(prefix="/learner", tags=["learner"])


@router.get("/stats")
async def get_learner_stats(token: str = Query(...)):
    payload = await decode_token_active(token)
    if not payload:
        raise HTTPException(status_code=401, detail="Token 無效")

    user_id: str = payload["sub"]
    concepts = await longterm_memory.get_all_concept_mastery(user_id)

    misconceptions = []
    for c in concepts:
        for p in c["confusion_patterns"]:
            if isinstance(p, dict):
                misconceptions.append({
                    "concept_name": c["concept_name"],
                    "pattern": p.get("pattern", ""),
                    "severity": p.get("severity", "medium"),
                    "repair_strategy": p.get("repair_strategy", ""),
                })
            elif isinstance(p, str) and p:
                misconceptions.append({
                    "concept_name": c["concept_name"],
                    "pattern": p,
                    "severity": "medium",
                    "repair_strategy": "",
                })

    return {
        "concepts": [
            {
                "concept_name": c["concept_name"],
                "mastery_score": c["mastery_score"],
                "total_exposures": c["total_exposures"],
                "last_tested": c["last_tested"],
            }
            for c in concepts
        ],
        "misconceptions": misconceptions,
        "weak_count": sum(1 for c in concepts if c["mastery_score"] < 0.6),
    }
